File: C++/459. Repeated Substring Pattern.cpp
class Solution {
public:
    bool repeatedSubstringPattern(string s) {
        vector<int> nex(s.length()+1, -1);
        int k = -1, i = 0;
        while(i < s.length())
        {
            if(k == -1 || s[i] == s[k])
            {
                i++, k++;
                nex[i] = k;
            }
            else
                k = nex[k];
        }
        int len = s.length();
        return nex[len] && nex[len] % (len-nex[len]) == 0;
    }
};

File: Python/459. Repeated Substring Pattern.py
class Solution:
    def repeatedSubstringPattern(self, s):
        """
        :type s: str
        :rtype: bool
        """
        nex = [-1] * (len(s) + 1)
        i, k = 0, -1
        while i < len(s):
            if k == -1 or s[i] == s[k]:
                i += 1
                k += 1
                nex[i] = k
            else:
                k = nex[k]
        return nex[len(s)] != 0 and nex[len(s)] % (len(s) - nex[len(s)]) == 0
